### PR TITLE
event and bill packet refactor

### PR DESCRIPTION
**Pull request by fgregg**
## Overview

This PR updates the way that we create Event and Bill packets.

## Testing Instructions

Assuming that you already have a local repo running

```bash
docker-compose run --rm -e DJANGO_MANAGEPY_MIGRATE=off app python manage.py compile_pdfs
```

Check out bill pages and event pages to see if the packets exist as expected. Not that it may take a moment for the packets to be processed by the pdf merger service.

Rerun

```bash
docker-compose run --rm -e DJANGO_MANAGEPY_MIGRATE=off app python manage.py compile_pdfs
```

No packets should be updated

If some time has passed since your last scrape, rerun a scrape and then rerun compile pdfs and any updated events and bills should update.

### CODE REVIEW

File: lametro/templates/partials/related_bills.html [hunk 1]
```diff
@@ -18,7 +18,7 @@ <h4>Board Reports</h4>
         <td>
           <a href='/board-report/{{ associated_bill.slug }}/' target="_blank">View</a>
         </td>
-        <td><a href={% if report.packet_url %}"{{report.packet_url}}"{% else %}"{{report.url}}"{% endif %}>Download</a></td>
+        <td><a href={% if report.packet.url %}"{{report.packet.url}}"{% else %}"{{report.url}}"{% endif %}>Download</a></td>
```
Comment (fgregg) [inline lametro/templates/partials/related_bills.html]: i don't know what the intention of "report.url" was. In django-councilmatic 1.0, bills doe not have a url attributes nor is one defined in la-metro.

File: lametro/models.py [hunk 2]
```diff
@@ -563,3 +563,115 @@ class Meta:
 
     guid = models.CharField(max_length=256)
     name = models.CharField(max_length=256, unique=True)
+
+
+class BillPacket(models.Model):
```
Comment (hancush) [inline lametro/models.py]: Seems like these `Packet` classes have some reusable attributes and logic. What about factoring that out into [Django's flavor of abstract base class](https://docs.djangoproject.com/en/2.2/topics/db/models/#abstract-base-classes)? Something like...

```python
class Packet(models.Model):

    class Meta:
        abstract = True

    updated_at = models.DateTimeField(auto_now=True)
    url = models.URLField()
    ready = models.BooleanField(default=False)

    @property
    def related_entity(self):
        raise NotImplementedError()

    @property
    def related_files(self):
        raise NotImplementedError()

    def save():
        self._merge_docs()
        self.url = settings.MERGER_BASE_URL + '/document/' + getattr(self, self.related_entity).slug
        super().save(*args, **kwargs)

    def is_ready(self):
        if not self.ready:
            response = requests.head(self.url)
            if response.status_code == 200:
                self.ready = True
                super().save()

        return self.ready

     def _merge_docs(self):
        merge_url = settings.MERGER_BASE_URL + '/merge_pdfs/' + getattr(self, self.related_entity).slug
        requests.post(merge_url, json=self.related_files)


class BillPacket(Packet):

    bill = models.OneToOneField(LAMetroBill,
                                related_name='packet',
                                on_delete=models.CASCADE)

    @property
    def related_entity(self):
        return self.bill

    @propert
    def related_files(self):
        ...
```

File: lametro/models.py [hunk 3]
```diff
@@ -563,3 +563,115 @@ class Meta:
...
+
+        self._merge_docs()
+
+        self.url = settings.MERGER_BASE_URL + '/document/' + self.bill.slug
+
+        response = requests.head(self.url)
+
+        super().save(*args, **kwargs)
+
+    def is_ready(self):
```
Comment (hancush) [inline lametro/models.py]: Where is this used?

File: lametro/management/commands/compile_pdfs.py [hunk 4]
```diff
@@ -1,233 +1,136 @@
...
+        return events
+
+    def _compile_board_reports(self):
+
+        bills = self._board_reports_to_compile()
+
+        for bill in tqdm.tqdm(bills, desc='Board Reports'):
+            bill_packet, created = BillPacket.objects.get_or_create(bill=bill)
+
+            # The docs in an bill packet are updated on save.
```
[the remark this answers is not in the document]
Comment (hancush) [reply lametro/management/commands/compile_pdfs.py]: Thanks for this docstring! Very helpful explanation.

File: lametro/management/commands/compile_pdfs.py [hunk 5]
```diff
@@ -1,233 +1,136 @@
...
+            # if the doc was just created, it was also saved so
+            # only save if the packet already existed
```
Comment (hancush) [inline lametro/management/commands/compile_pdfs.py]:
```suggestion
            # only call save here if the packet already existed
```

File: lametro/templates/partials/related_bills.html (hunk 1 above)
Comment (hancush) [reply lametro/templates/partials/related_bills.html]: Looks like it was meant to display document links if a report does not have an associated packet: https://github.com/datamade/la-metro-councilmatic/pull/271

File: lametro/management/commands/compile_pdfs.py [hunk 6]
```diff
@@ -1,233 +1,136 @@
...
-import logging
-import requests
-import json
-from datetime import datetime
-
 from django.core.management.base import BaseCommand
-from django.conf import settings
-
-from councilmatic.settings import MERGER_BASE_URL
+from django.db.models import F
```
Comment (hancush) [inline lametro/management/commands/compile_pdfs.py]: TIL!

File: lametro/management/commands/compile_pdfs.py [hunk 7]
```diff
@@ -1,233 +1,136 @@
...
 
 class Command(BaseCommand):
 
-    help = "This command compiles PDF packets for LA Metro events and board reports. By default, the command queries the new_billdocument table."
-
+    help = "This command compiles PDF packets for LA Metro events and board reports."
 
     def add_arguments(self, parser):
 
+        group = parser.add_mutually_exclusive_group(required=False)
```
[the remark this answers is not in the document]
Comment (fgregg) [reply lametro/management/commands/compile_pdfs.py]: no.

File: lametro/models.py (hunk 2 above)
Comment (fgregg) [reply lametro/models.py]: yeah, I went back and forth on this, but I think you are right that this is a good idea.

File: lametro/models.py [hunk 8]
```diff
@@ -563,3 +563,115 @@ class Meta:
...
+    url = models.URLField()
+    ready = models.BooleanField(default=False)
+
+    def save(self, *args, **kwargs):
+
+        self._merge_docs()
+
+        self.url = settings.MERGER_BASE_URL + '/document/' + self.bill.slug
+
+        response = requests.head(self.url)
```
[the remark this answers is not in the document]
Comment (fgregg) [reply lametro/models.py]: nope, leftover.

File: lametro/models.py (hunk 3 above)
Comment (fgregg) [reply lametro/models.py]: https://github.com/datamade/la-metro-councilmatic/blob/2.5/lametro/templates/lametro/event.html#L86

Comment (fgregg) [reply lametro/models.py]: https://github.com/datamade/la-metro-councilmatic/blob/2.5/lametro/templates/lametro/legislation.html

File: lametro/models.py [hunk 9]
```diff
@@ -563,3 +563,115 @@ class Meta:
...
+        for item in agenda_items:
+            for entity in item.related_entities.filter(bill__isnull=False):
+                bill_packet = BillPacket(bill=entity.bill)
+                related.extend(bill_packet.related_files)
+
+        return related
+
+    def _merge_docs(self):
+
+        merge_url = settings.MERGER_BASE_URL + '/merge_pdfs/' + self.event.slug
```
[the remark this answers is not in the document]
Comment (fgregg) [reply lametro/models.py]: oh, the joys of splitting up a pull request.

File: lametro/management/commands/compile_pdfs.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hancush) [reply lametro/management/commands/compile_pdfs.py]: Ok, thanks!

**Comment by fgregg** [comment]
hi @hancush should have responded to your review.

**Comment by hancush** [comment]
Almost everything is looking and working great, @fgregg! The only hiccup is I'm still not seeing the packet link for board reports. Compare:

- http://localhost:8011/board-report/2019-0679/
- https://boardagendas.metro.net/board-report/2019-0679/

I think it may actually be a data issue: It doesn't look like we're getting all (any?) of the bill attachments in the scrape, which I re-ran just to be sure. Is this expected? (Seems unrelated to this PR, so happy to have that addressed in a different PR, if it's not an error with my local setup!)

**Comment by fgregg** [comment]
hi @hancush, very sorry for neglecting to say this, but this was something I had to fix upstream in django-councilmatic.

You'll need to rebuild the app container, and then it should work.

**Comment by hancush** [comment]
Right on, thanks @fgregg!

**Pull request merged**